### PR TITLE
Refactor of evented runners (#17283)

### PR DESCRIPTION
**Pull request by jacksontj**
This is a refactor of #17283 (tangentially related to #9444 since we had some discussion in there) to avoid most of the code duplication-- since we now have async client mixins.

I'm still not sure about the **progress** global, since it is _very_ specific, but I'm not sure having someone having to write out something like:

```
__jid_event__.fire_event({'data': 'Runner is {0}% done'.format(i), 'outputter': 'pprint'}, 'progress')
```

Is any better than just 

```
event.fire_event({'data': 'Runner is {0}% done'.format(i), 'outputter': 'pprint'}, 'progress', base=__tag__)
```

So in addition to the **progress** global I've added **jid**, **user**, **tag**, and **jid_event**.

As part of that the "low" func now creates a JID and events for all calls to it (so we are consistent). All of these are specific to each execution. **jid_event** is a new class called "NamespacedEvent" which is just a wrapper to allow for people to fire events from a different base tag, in this case within the runner you could fire an event with tag "foo" and it would end up as "salt/run/SOME_JID/foo" (alternatively you could just use base=**tag** with the regular event class (which I'm leaning towards, since this seems a bit messy)

@cachedout @whiteinge 
Pinging the 2 of you about the **progress** object vs the Namespaced event. I'm thinking we could maybe generalize the outputter class in the runner to work for wheel and runner, and come up with some way of printing out all events within the namespace-- so people don't have to follow the strict event format from RunnerEvent (which I've removed).

### CODE REVIEW

**Comment by cachedout** [comment]
I'm fine with going the namespaced direction, so long as we update the docs and such accordingly. I think it's more flexible and all around a better approach then progress events.

**Comment by jacksontj** [comment]
@cachedout K, I'll get a cleanup of that once I finish fixing/cleaning all the arg/kwarg parsing stuff. There are at least 3 different ways for the function calls to go and they all seem to parse it differently... so I'm just un-tangling ;)

**Comment by cachedout** [comment]
Sounds great! I'm looking forward to seeing the results.

**Comment by jacksontj** [comment]
Wheel uses "low" as kwargs, and runners seem to actually use it as a lowstate object-- so I'm going to have to make them the same-- since they have the same mixins under the hood. :/

**Comment by cachedout** [comment]
I think that should work. I'm going to ping @whiteinge since he might want to chime in here as well.

**Comment by whiteinge** [comment]
Agree with Thomas Jackson on namespaced direction. Also right about
lowstate diff between runner and wheel; I've had to work around that before
too.

**Comment by jacksontj** [comment]
@whiteinge @cachedout So, I think I have it. The "lowstate" object thing that is doing is a bit odd. Instead of 'args' and "kwargs' as parts of the low object, all kwargs are basically just merged into low. I don't *think* thats what we do elsewhere, but I can't remember off hand.

**Comment by jacksontj** [comment]
After some quick looking that way (kwargs as part of the main dict) matches the pub data structures-- so that was probably intentional, albeit confusing.

**Comment by whiteinge** [comment]
Yeah, think you're right about that. Intentional but confusing.

**Comment by jacksontj** [comment]
Okay, I *think* that covers all my bases. I'll wait for the test runs to finish, but that last batch includes a *lot* of cleanup. RunnerClient and Wheel Client are basically the same-- except a couple little things. So instead of maintaining 2 completely separate classes (as we have) I've pushed even more into the mixins. This means that the interface to the wheel cmd() calls changed *slightly*, but it was kind of broken before (couldn't pass args-- for example). 

I still need to migrate the rest of the __progress__ instances to the event firing (should be simple). The only other thing I'd like to do is have the runner die if you control-c out of a synchronous call on the command line-- since that would make it backwards compatible, unless we don't care...

**Comment by jacksontj** [comment]
As I'm migrating these, I wonder if we should just only print a subset of suffixes (such as new, progress, and ret) that way if you want additional events they wont *have* to be printed, and we wont have as much duplication (right now it prints tag: dict-- which is usually something like progress: data: the message here, where we could have just the dict).

**Comment by whiteinge** [comment]
I like that idea.

**Comment by jacksontj** [comment]
More weirdness, we actually have some tests (https://github.com/jacksontj/salt/blob/develop/tests/integration/client/runner.py#L74) which enforce the kwargs as part of the top level "low" object, just to make it more confusing

**Comment by jacksontj** [comment]
@whiteinge which tag name should we use? Want to stick with "progress" or should we move to something more generic, like "print" or "out"? I overloaded the "print" function to fire events to the tag "print", but we can support multiple tags if we want-- or make them use the same ones.

**Comment by jacksontj** [comment]
Looks like the VMs didn't start on jenkins-- closed + re-opened to trigger a new build

**Comment by jacksontj** [comment]
@rallytime Tests have been running for ~12 hours now, so I think its safe to say they won't ever finish ;)

**Comment by jacksontj** [comment]
@rallytime Looks like the linode account ran out of cash:

```
23:35:14 (8) Unable to add a new Linode as your account has an outstanding balance. Please submit a payment, or wait for us to automatically charge your card.
```

**Comment by cachedout** [comment]
@jacksontj Looks like a CC expired at the end of the year. Oops! We'll get it sorted ASAP.

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
Our Linode account should be fixed, I updated the card

**Comment by jacksontj** [comment]
@rallytime Jenkins seems to be dead. I have a few open PRs and I can't seem to get it to run for any of them :(

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by jacksontj** [comment]
@rallytime @thatch45 And that they are! After the last cleanup of printing out the events I seemed to have accidentally deleted the code that set the ret. Its re-added now

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
We seriously need Jenkins to be more reliable.....

**Comment by jacksontj** [comment]
@thatch45  +1

Might I suggest by starting with some metrics/alerting tracking how its performing? Easier to fix reliability if you have metrics :)

**Comment by thatch45** [comment]
Looks like these failures are legit, the wheel system is closely related to runners but the test results are missing from Jenkins....

**Comment by jacksontj** [comment]
Yea, the few that I've looked at are legitimate failures, but not legitimate tests. For example,  the integration.runners.manage.ManageTest.test_up test is calling the runner with --async and expecting the return on the CLI (d746bb2). This was because the old implementation "async" wasn't really async ;)

**Comment by thatch45** [comment]
Thanks!

**Comment by jacksontj** [comment]
I think that covers them all-- but lets wait for the test runs to finish, I don't want to accidentally break backwards compat on something ;)

**Comment by jacksontj** [comment]
@rallytime looks like the Ubuntu VM failed immediately :(

**Comment by rallytime** [comment]
Yes, we are having trouble with Jenkins again. We're woking on getting it more stable as we speak. :)

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Hrm. @jacksontj Now that we actually have the tests running here, it looks like both VMs are hanging right after this test:

test_cmd_sync (integration.client.runner.RunnerModuleTest) ... OK (0.052s)

I haven't seen this on other branches, so I think this might need to be legitimately addressed.

**Comment by rallytime** [comment]
I ended up aborting the tests since they were hanging. Seems GH still things they're running in yellow, but they are in fact aborted where only a subset of the tests were finished before aborting.

**Comment by jacksontj** [comment]
@rallytime I ran integration.client.runner.RunnerModuleTest (all 6 of them) locally and it passed without issue on my box :/

**Comment by rallytime** [comment]
Ok. Jenkins is kind of backed up right now and has a pretty hefty queue. I'll circle back around to this and take another look and re-trigger these to see if it will happen again. Thanks again for your patience!

**Comment by jacksontj** [comment]
Sure, no problem :)

**Comment by jacksontj** [comment]
Still need to update the docs before we merge this, but I want to see if we can get a test run completed :)

**Comment by jacksontj** [comment]
Can't reproduce that test failure locally.. hopefully it was just a fluke

**Comment by rallytime** [comment]
I haven't seen these failures on the develop branch, and we've got it pretty stable right now, so these are likely legitimate. Especially since one of the tests is failing on both test VMs. Is this an older version of develop, or has it been rebased within the last day or so? I'll see if I can reproduce these as well.

**Comment by jacksontj** [comment]
Its a fairly old develop, so who knows. Since I was unable to repro locally I just rebased against current develop.

**Comment by jacksontj** [comment]
@rallytime  Looks like a Jenkins rough patch... Getting all sorts of java errors :(

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
@jacksontj These failures are consistent here, as the exact same runner tests are failing as before and they're not failing in the develop branch, so they're not left over from something else. As to why they're failing, I'll have to take a close look. :)

**Comment by jacksontj** [comment]
Those test failures should be due a change from a while ago where @cachedout had made *all* runner tests async-- which messed up some of these. I can't get the test to fail locally, but I also had some *major* rebase conflicts last attempt-- so its possible i completely messed it up ;) I just pulled in another fix for this stuff and we'll see how this test run goes.

**Comment by jacksontj** [comment]
Another big rebase. So the failures weren't terribly hard to fix. The wheel ones are because the wheel interface is inconsistent with the runner ones (added TODO markers to clean it up).

The runner one was actually just exposing a race condition in the SaltEvent class. ZMQ connect() doesn't guarantee that the underlying socket is actually connected. I put in a workaround for this, but AFAIK there isn't a good fix for that.

**Comment by jacksontj** [comment]
Look like the test failures are just timeouts of tornado stuff, everything else seems fine. Those timeouts usually mean Jenkins is busy

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by jacksontj** [comment]
I'd like to get this one wrapped up this week, as we *need* to get this into 2015.2, as the current runner __progress__ implementation breaks backwards compatibility and exposes an interface we don't want to support.

**Comment by thatch45** [comment]
Yes, we completely agree, I think that this is ready to merge

**Comment by thatch45** [comment]
Ok, merged!
We will need to watch the develop tests for these issues and get them cleaned up ASAP @jacksontj so that we can backport this to 2015.2

**Comment by jacksontj** [comment]
@cachedout Would you be able to change the docs over to this new markup? I can check it after you finish-- probably a good way to get some additional code review done :)

**Comment by jacksontj** [comment]
The only other thing I'd like to get in is some way to stop runners when they aren't run in async-- that way it would be backwards compatible.

**Comment by cachedout** [comment]
@jacksontj Definitely. I'll do that now.

**Comment by jacksontj** [comment]
@cachedout Any update on docs? I've actually had a couple people ping me about this :) I think I'm going to make one more change to this tomorrow to have the non-async runners execute in a thread instead of a daemonized process (so stdout print will work and so you can stop the runner execution).

**Comment by cachedout** [comment]
@jacksontj Got distracted with some other fires. :] Will get back to docs today. Also, @kiorky has some feedback on the runner changes in #19707 if you have a moment.

File: tests/integration/utils/test_reactor.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            fh.write('''
+update_fileserver:
+  runner.fileserver.update
+''')
+
+    def tearDown(self):
+        if os.path.isdir(self.tempdir):
+            shutil.rmtree(self.tempdir)
+
+    def test_basic(self):
```
Comment (s0undt3ch) [inline tests/integration/utils/test_reactor.py]: What is this supposed to test?

**Pull request merged**